File: tema5_cross_platform/server_csharp/musicFestival/protobuf/ProtoWorker.cs
using System;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Threading;
using csharpMusicFestival.validator;
using Google.Protobuf;
using Services;

namespace Proto
{
    public class ProtoWorker : IObserver
    {
		private IService server;
		private TcpClient connection;
		private NetworkStream stream;
		private IFormatter formatter;
		private volatile bool connected;

		public ProtoWorker(IService server, TcpClient connection)
		{
			this.server = server;
			this.connection = connection;
			try
			{
				stream = connection.GetStream();
				formatter = new BinaryFormatter();
				connected = true;
			}
			catch (Exception e)
			{
				Console.WriteLine(e.StackTrace);
			}
		}
		
		public virtual void run()
		{
			while(connected)
			{
				try
				{
					Request request = Request.Parser.ParseDelimitedFrom(stream);
					Response response =handleRequest(request);
					if (response!=null)
					{
						sendResponse(response);
					}
				}
				catch (Exception e)
				{
					Console.WriteLine(e.StackTrace);
				}
				
				try
				{
					Thread.Sleep(1000);
				}
				catch (Exception e)
				{
					Console.WriteLine(e.StackTrace);
				}
			}
			try
			{
				stream.Close();
				connection.Close();
			}
			catch (Exception e)
			{
				Console.WriteLine("Error "+e);
			}
		}

		private void sendResponse(Response response)
		{
			Console.WriteLine("sending response "+response);
			lock (stream)
			{
				response.WriteDelimitedTo(stream);
				stream.Flush();
			}

		}

		private Response handleRequest(Request request)
		{
			Response response = null;
			Request.Types.Type reqType = request.Type;
			switch (reqType)
			{
				case Request.Types.Type.Login:
				{
					Console.WriteLine("Login request ...");
					csharpMusicFestival.domain.User user = ProtoUtils.GetUser(request);
					try
					{
						lock (server)
						{
							server.Login(user, this);
						}

						return ProtoUtils.createOkResponse();
					}
					catch (Error e)
					{
						connected = false;
						return ProtoUtils.createErrorResponse(e.Message);
					}
				}

				case Request.Types.Type.Logout:
				{
					Console.WriteLine("Logout request");
					csharpMusicFestival.domain.User user = ProtoUtils.GetUser(request);
					try
					{
						lock (server)
						{
							server.Logout(user, this);
						}

						connected = false;
						return ProtoUtils.createOkResponse();

					}
					catch (Error e)
					{
						return ProtoUtils.createErrorResponse(e.Message);
					}
				}

				case Request.Types.Type.GetShows:
				{
					Console.WriteLine("Get shows request");
					try
					{
						csharpMusicFestival.domain.Show[] shows = server.GetAll();
						return ProtoUtils.CreateGetShowsResponse(shows);
					}
					catch (Error e)
					{
						return ProtoUtils.createErrorResponse(e.Message);
					}
				}

				case Request.Types.Type.GetArtistsByDate:
				{
					Console.WriteLine("Get artists by date request");
					String date = request.Date;
					try
					{
						csharpMusicFestival.domain.Artist[] artists = server.GetArtists(date);
						return ProtoUtils.CreateGetArtistsByDateResponse(artists);
					}
					catch (Error e)
					{
						return ProtoUtils.createErrorResponse(e.Message);
					}
				}

				case Request.Types.Type.BuyTicket:
				{
					Console.WriteLine("Buy ticket request");
					csharpMusicFestival.domain.Ticket ticket = ProtoUtils.GetTicket(request);
					try
					{
						lock (server)
						{
							server.SellTickets(ticket);
						}

						return ProtoUtils.createOkResponse();
					}
					catch (InvalidPurchaseException e)
					{
						return ProtoUtils.createErrorResponse(e.Message);
					}
					catch (Error e)
					{
						return ProtoUtils.createErrorResponse(e.Message);
					}
				}
					
			}
			return response;
		}

		public void TicketSold(csharpMusicFestival.domain.Ticket ticket)
		{
			Console.WriteLine("Ticket sold  " + ticket);
			try
			{
				sendResponse(ProtoUtils.CreateBuyTicketResponse(ticket));
			}
			catch (Exception e)
			{
				throw new Error("Sending error: " + e);
			}
		}
    }
}

File: tema5_cross_platform/client_java/Networking/src/main/java/ProtoWorker.java
import domain.Artist;
import domain.Show;
import domain.Ticket;
import domain.User;
import objectprotocol.ErrorResponse;
import objectprotocol.GetArtistsByDateRequest;
import protobuffprotocol.Protobufs;
import validator.InvalidPurchaseException;
import java.io.*;
import java.net.Socket;


public class ProtoWorker implements Runnable, IObserver {
    private IService server;
    private Socket connection;
    private InputStream input;
    private OutputStream output;
    private volatile boolean connected;

    public ProtoWorker(IService server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try{
            output=connection.getOutputStream() ;//new ObjectOutputStream(connection.getOutputStream());
            input=connection.getInputStream(); //new ObjectInputStream(connection.getInputStream());
            connected=true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void run() {
        while(connected){
            try {
                // Object request=input.readObject();
                System.out.println("Waiting requests ...");
                Protobufs.Request request = Protobufs.Request.parseDelimitedFrom(input);
                System.out.println("Request received: "+request);
                Protobufs.Response response = (Protobufs.Response) handleRequest(request);
                if (response!=null){
                    sendResponse(response);
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            System.out.println("Error "+e);
        }
    }

    private Serializable handleRequest(Protobufs.Request request) {
        Protobufs.Response response = null;
        switch (request.getType()) {
            case Login: {
                System.out.println("Login request ...");
                User user = ProtoUtils.getUser(request);
                try {
                    server.login(user, this);
                    return ProtoUtils.createOkResponse();
                } catch (Error e) {
                    connected=false;
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case Logout:{
                System.out.println("Logout request");
                User user=ProtoUtils.getUser(request);
                try {
                    server.logout(user, this);
                    connected=false;
                    return ProtoUtils.createOkResponse();

                } catch (Error e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case GetShows: {
                System.out.println("Get shows request");
                try {
                    Show[] shows = server.getAll();
                    return ProtoUtils.createGetShowsResponse(shows);
                } catch (Error e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case GetArtistsByDate: {
                System.out.println("Get artists by date request");
                String date = ProtoUtils.getDate(request);
                try {
                    Artist[] artists = server.getArtists(date);
                    return ProtoUtils.createGetArtistsByDateResponse(artists);
                } catch (Error e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }

            case BuyTicket: {
                System.out.println("Buy ticket request");
                Ticket ticket = ProtoUtils.getTicket(request);
                try {
                    server.sellTickets(ticket);
                    return ProtoUtils.createOkResponse();
                } catch (Error | InvalidPurchaseException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
        }

        return response;
    }

    private void sendResponse(Protobufs.Response response) throws IOException{
        System.out.println("sending response "+response);
        response.writeDelimitedTo(output);
        //output.writeObject(response);
        output.flush();
    }

    @Override
    public void ticketSold(Ticket ticket) throws Error {
        System.out.println("Ticket sold  "+ ticket);
        try {
            sendResponse(ProtoUtils.createTicketSoldResponse(ticket));
        } catch (IOException e) {
            throw new Error("Sending error: "+e);
        }
    }
}
